### PR TITLE
Ce 2125 approve action

### PR DESCRIPTION
**Pull request by kamilkoterba**
https://wikia-inc.atlassian.net/browse/CE-2125

Added ?action=approvedraft and right rail module with approval button
Also renamed naming of first module as now we have two: for creating draft and for approving

@Wikia/community-engineering

### CODE REVIEW

File: includes/DefaultSettings.php [hunk 1]
```diff
@@ -5335,6 +5335,7 @@
  * Unsetting core actions will probably cause things to complain loudly.
  */
 $wgActions = array(
+	'approvedraft'   => true,
```
Comment (lgarczewski) [inline includes/DefaultSettings.php]: This shouldn't be here. Instead, please add to $wgActions in the code of our extensions, similarly to how we add items to $wgHooks and other globals.

File: includes/AutoLoader.php [hunk 2]
```diff
@@ -258,6 +258,7 @@
 	'ZipDirectoryReader' => 'includes/ZipDirectoryReader.php',
 
 	# includes/actions
+	'ApprovedraftAction' => 'includes/actions/ApprovedraftAction.php',
```
Comment (lgarczewski) [inline includes/AutoLoader.php]: This should not be here. Please add via $wgAutoLoadClasses in the extension setup file.

File: includes/actions/ApprovedraftAction.php [hunk 3]
```diff
@@ -0,0 +1,50 @@
+<?php
+/**
+ * Move draft page content to parent page
```
Comment (lgarczewski) [inline includes/actions/ApprovedraftAction.php]: This file should live with the TemplateDraft extension, not in includes.

File: skins/oasis/modules/TemplateDraftModuleController.class.php [hunk 4]
```diff
@@ -14,4 +18,20 @@ public function executeIndex( $params ) {
...
+	public function executeApprove( $params ) {
+		$this->wg->Out->addStyle(
+			AssetsManager::getInstance()->getSassCommonURL( "skins/oasis/css/modules/TemplateDraftModule.scss" )
+		);
+
+		/**
+		 * TODO: Make a better URL composing script
+		 */
+		$titleUrl = $this->app->wg->Title->getFullURL();
+		$this->draftUrl = "{$titleUrl}?action=approvedraft";
```
Comment (lgarczewski) [inline skins/oasis/modules/TemplateDraftModuleController.class.php]: Please use:
```php
$title->getFullUrl( ['action'] => 'approvedraft'] );
```

File: skins/oasis/modules/templates/TemplateDraftModule_Create.php [hunk 5]
```diff
@@ -0,0 +1,13 @@
+<section class="templatedraft-module module">
+	<h2><?= wfMessage( 'templatedraft-module-title-create' )->escaped() ?></h2>
+	<p class="templatedraft-module-subtitle"><?= wfMessage( 'templatedraft-module-subtitle-create' )->parse() ?></p>
```
Comment (lgarczewski) [inline skins/oasis/modules/templates/TemplateDraftModule_Create.php]: Doesn't `wfMessage::parse()` add `<p>` tags automatically?

Comment (Grunny) [inline skins/oasis/modules/templates/TemplateDraftModule_Create.php]: @lgarczewski Nope, `wfMessage()->parse()` parses inline, while `wfMessage()->parseAsBlock()` wraps the output in `<p></p>`.

Comment (kamilkoterba) [inline skins/oasis/modules/templates/TemplateDraftModule_Create.php]: and we need class here which is not doable with parseAsBlock right?

File: extensions/wikia/TemplateDraft/ApprovedraftAction.php [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+	public function getName() {
+		return 'approvedraft';
+	}
+
+	protected function getDescription() {
+		return '';
+	}
+
+	public function onView() {
+		global $wgTitle;
```
Comment (Grunny) [inline extensions/wikia/TemplateDraft/ApprovedraftAction.php]: Action has access to the request context, so you can just use `$this->getTitle()` to get the title and avoid the global.

File: extensions/wikia/TemplateDraft/ApprovedraftAction.php [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+	}
+
+	protected function getDescription() {
+		return '';
+	}
+
+	public function onView() {
+		$title = $this->getTitle();
+
+		$this->redirectParams = wfArrayToCGI( array_diff_key(
```
Comment (Grunny) [inline extensions/wikia/TemplateDraft/ApprovedraftAction.php]: Why isn't this just a local variable?

File: extensions/wikia/TemplateDraft/ApprovedraftAction.php [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+			 * Show a friendly error message to a user after redirect
+			 */
+			BannerNotificationsController::addConfirmation(
+				wfMessage( 'templatedraft-approval-no-page-error' )->escaped(),
+				BannerNotificationsController::CONFIRMATION_ERROR
+			);
+
+			$this->getOutput()->redirect( $title->getFullUrl( $this->redirectParams ) );
+		} else {
+			$this->redirectTitle = $title->getBaseText();
```
Comment (Grunny) [inline extensions/wikia/TemplateDraft/ApprovedraftAction.php]: Likewise, why is this a class property?

File: extensions/wikia/TemplateDraft/ApprovedraftAction.php [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+			$this->redirectTitle = Title::newFromText( $this->redirectTitle, $title->getNamespace() );
+			$templateDraftHelper = new TemplateDraftHelper();
+			$templateDraftHelper->approveDraft( $title );
```
Comment (Grunny) [inline extensions/wikia/TemplateDraft/ApprovedraftAction.php]: Should we verify `$title` is both a template and a draft page (i.e. `TemplateDraftHelper::isTitleDraft`)? Right now it doesn't check, but `approveDraft` will take the `baseText` and create the title in template namespace even if it wasn't originally due to it being used on a non-template or draft page.

File: extensions/wikia/TemplateDraft/TemplateDraft.setup.php [hunk 10]
```diff
@@ -57,3 +57,9 @@
  * Right rail module
  */
 $wgAutoloadClasses['TemplateDraftModuleController'] = $IP . '/skins/oasis/modules/TemplateDraftModuleController.class.php';
+
+/**
+ * Add approvedraft action (?action=apprevedraft)
+ */
+$wgAutoloadLocalClasses['ApprovedraftAction'] = 'extensions/wikia/TemplateDraft/ApprovedraftAction.php';
```
Comment (Grunny) [inline extensions/wikia/TemplateDraft/TemplateDraft.setup.php]: Should be `$wgAutoloadClasses['ApprovedraftAction'] = __DIR__ . '/ApprovedraftAction.php';` as normal.

File: extensions/wikia/TemplateDraft/ApprovedraftAction.php [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+		$title = $this->getTitle();
+
+		$redirectParams = wfArrayToCGI( array_diff_key(
+			$this->getRequest()->getQueryValues(),
+			[ 'title' => null, 'action' => null ]
+		));
+
+		$templateDraftHelper = new TemplateDraftHelper();
+
+		if ( !$title->exists() ) {
```
Comment (lukaszkonieczny) [inline extensions/wikia/TemplateDraft/ApprovedraftAction.php]: Maybe we can use ```if ( !$title->exists() || !$templateDraftHelper->isTitleDraft( $title ) )``` while there is the same code in these two checks

**Comment by lukaszkonieczny** [comment]
lgtm

**Pull request merged**